controllers/storagecluster/server.go

```diff
@@ -0,0 +1,190 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/multierr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+)
+
+var (
+	ServerImage = "quay.io/sp1098/ocs-operator:consumer-api"
+)
+
+type ocsProviderServer struct{}
+
+func (obj *ocsProviderServer) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	if !instance.Spec.AllowRemoteStorageConsumers {
+		r.Log.Info("nigoyal Spec.AllowRemoteStorageConsumers is disabled")
+		return obj.ensureDeleted(r, instance)
+	}
+
+	r.Log.Info("nigoyal Spec.AllowRemoteStorageConsumers is enabled creating resources")
+
+	desiredServerService := GetProviderServerService(instance)
+	desiredServerDeployment := GetProviderServerDeployment(instance)
+
+	var finalErr error
+
+	for _, obj := range []client.Object{
+		desiredServerDeployment,
+		desiredServerService,
+	} {
+		_, err := controllerutil.CreateOrUpdate(
+			context.TODO(), r.Client, obj,
+			func() error { return nil },
```

Comment:
I might be forgetting, but I believe the `obj` being passed here would be overridden by `CreateOrUpdate()` with whatever it finds (or doesn't find). See the example here, where the desired PodTemplateSpec is defined separately and outside the `deploy` object: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate Thus, you must use this in-line function to enforce the state you want.

Reply:
We use this pattern a lot in the deployer. 
The behaviour is a little bit different"
The object is taken then the result of the get is MERGED into it then the callback is run against the merged object to allow programmatic changes to the content before the create/update call is issued on the result.

The key point is that the merge operation will only affect fields that are defined in the k8s resource. This means that if the object presented to the function contains values that are undefined in the resource the outcome will be a struct that contains a merge of both.

Reply:
Sure. And that sounds like something we wouldn't want, right? If the Deployment we Get() has a different PodSpec, wouldn't that override the desired PodSpec you want to actually have?

Reply:
Usually, you start with an empty obj! you should not think of object passed as the template
For the deployer we do the following:
1. create an "empty" obj
2. run `createOrUpdate`
3. have a callback that takes the content of the template and replace/merge it into the obj

An example: https://github.com/red-hat-storage/ocs-osd-deployer/blob/main/controllers/managedocs_controller.go#L599-L626

Reply:
@jarrpa we are in agreement here! the mutate function is the one to enforce desired state

Reply:
Yes we are, I had an discussion with Jose on the same

Reply:
done